// _lesslint@1.0.4@lesslint/lib/rule/single-comment.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

'use strict';

/**
 * 规则名称
 *
 * @const
 * @type {string}
 */
/**
 * @file 注释检验
 *       单行注释尽量使用 // 方式
 *       https://github.com/ecomfe/spec/blob/master/less-code-style.md#%E6%B3%A8%E9%87%8A
 * @author ielgnaw(wuji0223@gmail.com)
 */

var RULENAME = 'single-comment';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (!opts.ruleVal) {
            return;
        }

        css.walkComments(function (comment) {
            // 排除单行注释
            if (comment.inline) {
                return;
            }

            var source = comment.source;


            var startLine = source.start.line;
            var endLine = source.end.line;
            if (startLine === endLine) {
                var lineContent = (0, _util.getLineContent)(startLine, opts.fileContent);

                result.warn(RULENAME, {
                    node: comment,
                    ruleName: RULENAME,
                    line: startLine,
                    col: source.start.column,
                    message: '`' + lineContent + '` Single Comment should be use `//`',
                    colorMessage: '`' + lineContent.replace(lineContent, function ($1) {
                        return _chalk2.default.magenta($1);
                    }) + '` ' + _chalk2.default.grey('Single Comment should be use `//`')
                });
            }
        });
    };
});
