// _lesslint@1.0.4@lesslint/lib/rule/variable-name.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

'use strict';

/**
 * 规则名称
 *
 * @const
 * @type {string}
 */
/**
 * @file 变量检验
 *       变量命名必须（MUST）采用 @foo-bar 形式，不得（MUST NOT）使用 @fooBar 形式。
 *       https://github.com/ecomfe/spec/blob/master/less-code-style.md#%E5%8F%98%E9%87%8F
 * @author ielgnaw(wuji0223@gmail.com)
 */

var RULENAME = 'variable-name';

/**
 * 匹配变量名字的正则
 *
 * @const
 * @type {RegExp}
 */
var VARIABLE_NAME_REG = /^@([a-z0-9\-]+)$/;

/**
 * 错误信息
 *
 * @const
 * @type {string}
 */
var MSG = 'Variable name must be like this `@foo-bar or @foobar`';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (!opts.ruleVal) {
            return;
        }

        css.walkDecls(function (decl) {
            var source = decl.source,
                prop = decl.prop;

            var lineNum = source.start.line;
            var lineContent = (0, _util.getLineContent)(lineNum, opts.fileContent);
            if (/^@/.test(prop) && !VARIABLE_NAME_REG.test(prop)) {
                result.warn(RULENAME, {
                    node: decl,
                    ruleName: RULENAME,
                    line: lineNum,
                    col: source.start.column,
                    message: '`' + lineContent + '` ' + MSG,
                    colorMessage: '`' + lineContent.replace(prop, function ($1) {
                        return _chalk2.default.magenta($1);
                    }) + '` ' + _chalk2.default.grey(MSG)
                });
            }
        });
    };
});
