// _csshint@0.3.4@csshint/lib/rule/always-semicolon.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'always-semicolon';

/**
 * 错误信息
 *
 * @const
 * @type {string}
 */
/**
 * @file always-semicolon 的检测逻辑
 *       012: [强制] 属性定义后必须以分号结尾。
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = 'Attribute definition must end with a semicolon';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (!opts.ruleVal) {
            return;
        }

        css.walkRules(function (rule) {
            if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                return;
            }

            if (rule.raws.semicolon) {
                return;
            }

            var lastProp = rule.nodes[rule.nodes.length - 1];
            if (lastProp && lastProp.type !== 'comment') {
                var source = lastProp.source;
                var line = source.start.line;
                var lineContent = (0, _util.getLineContent)(line, source.input.css) || '';

                var value = lastProp.important ? lastProp.value + (lastProp.important ? lastProp.important : ' !important') : lastProp.value;

                var colorStr = lastProp.prop + lastProp.raws.between + value;
                var col = source.start.column + colorStr.length;

                result.warn(RULENAME, {
                    node: rule,
                    ruleName: RULENAME,
                    line: line,
                    col: col,
                    message: MSG,
                    colorMessage: '`' + lineContent.replace(colorStr, _chalk2.default.magenta(colorStr)) + '` ' + _chalk2.default.grey(MSG)
                });
                global.CSSHINT_INVALID_ALL_COUNT++;
            }
        });
    };
});
