// _csshint@0.3.4@csshint/lib/rule/gradients.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'gradients';

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
/**
 * @file gradients 的检测逻辑
 *       When using a vendor-prefixed gradient, make sure to use them all
 *       https://github.com/CSSLint/csslint/wiki/Require-all-gradient-definitions
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = 'Missing vendor-prefixed CSS gradients for ';

var gradients = {};

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {
        if (!opts.ruleVal) {
            return;
        }

        css.walkRules(function (rule) {

            gradients = {
                moz: 0,
                webkit: 0,
                oldWebkit: 0,
                o: 0
            };

            rule.walkDecls(function (decl) {
                if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                    return;
                }

                var value = decl.value;
                if (/\-(moz|o|webkit)(?:\-(?:linear|radial))\-gradient/i.test(value)) {
                    gradients[RegExp.$1] = 1;
                } else if (/\-webkit\-gradient/i.test(value)) {
                    gradients.oldWebkit = 1;
                }
            });

            var missing = [];

            if (!gradients.moz) {
                missing.push('Firefox 3.6+: -moz-linear-gradient');
            }

            if (!gradients.webkit) {
                missing.push('Webkit (Safari 5+, Chrome): -webkit-linear-gradient');
            }

            if (!gradients.oldWebkit) {
                missing.push('Old Webkit (Safari 4+, Chrome): -webkit-gradient');
            }

            if (!gradients.o) {
                missing.push('Opera 11.1+: -o-linear-gradient');
            }

            if (missing.length && missing.length < 4) {
                var source = rule.source;
                var line = source.start.line;
                var lineContent = (0, _util.getLineContent)(line, source.input.css);
                var col = source.start.column;
                var str = MSG + missing.join(', ');
                result.warn(RULENAME, {
                    node: rule,
                    ruleName: RULENAME,
                    line: line,
                    col: col,
                    message: str,
                    colorMessage: '`' + lineContent.replace(rule.selector, _chalk2.default.magenta(rule.selector)) + '` ' + _chalk2.default.grey(str)
                });
                global.CSSHINT_INVALID_ALL_COUNT++;
            }
        });
    };
});
