// _csshint@0.3.4@csshint/lib/rule/regex-selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.check = undefined;

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _postcss = require('postcss');

var _postcss2 = _interopRequireDefault(_postcss);

var _util = require('../util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * 当前文件所代表的规则名称
 *
 * @const
 * @type {string}
 */
var RULENAME = 'regex-selectors';

/**
 * 错误的信息
 *
 * @const
 * @type {string}
 */
/**
 * @file regex-selectors 的检测逻辑
 *       Selectors that look like regular expressions are slow and should be avoided
 *       https://github.com/CSSLint/csslint/wiki/Disallow-selectors-that-look-like-regular-expressions
 * @author ielgnaw(wuji0223@gmail.com)
 */

var MSG = 'Selectors that look like regular expressions are slow and should be avoided';

/**
 * 具体的检测逻辑
 *
 * @param {Object} opts 参数
 * @param {*} opts.ruleVal 当前规则具体配置的值
 * @param {string} opts.fileContent 文件内容
 * @param {string} opts.filePath 文件路径
 */
var check = exports.check = _postcss2.default.plugin(RULENAME, function (opts) {
    return function (css, result) {

        if (opts.ruleVal) {

            css.walkRules(function (rule) {
                if (global.CSSHINT_INVALID_ALL_COUNT >= opts.maxError) {
                    return;
                }

                var selector = rule.selector,
                    source = rule.source;


                var selectorGroup = selector.split(',');
                var line = source.start.line;
                var lineContent = (0, _util.getLineContent)(line, source.input.css);

                for (var i = 0, len = selectorGroup.length; i < len; i++) {
                    var selectorInGroup = selectorGroup[i] || '';
                    var attrs = selectorInGroup.match(/\[.+?\](?::[^\s>+~\.#\[]+)?/g);
                    if (!attrs) {
                        continue;
                    }

                    if (selectorInGroup.slice(0, 1) === '\n') {
                        line = line + 1;
                        lineContent = (0, _util.getLineContent)(line, source.input.css);
                    }

                    for (var j = 0, attrsLen = attrs.length; j < attrsLen; j++) {
                        var attr = attrs[j];
                        if (/([\~\|\^\$\*]=)/.test(attr)) {
                            var col = lineContent.indexOf(attr) + 1;
                            result.warn(RULENAME, {
                                node: rule,
                                ruleName: RULENAME,
                                line: line,
                                col: col,
                                message: MSG,
                                colorMessage: '`' + lineContent.replace(attr, _chalk2.default.magenta(attr)) + '` ' + _chalk2.default.grey(MSG)
                            });

                            global.CSSHINT_INVALID_ALL_COUNT++;
                        }
                    }
                }
            });
        }
    };
});
